packages/base-driver/lib/basedriver/helpers.js

```diff
@@ -66,73 +67,157 @@ async function retrieveHeaders (link) {
+}
+
+async function calculateFileIntegrity (filePath) {
+  return await fs.hash(filePath);
+}
+
+async function isAppIntegrityOk (currentPath, expectedIntegrity = {}) {
+  if (!await fs.exists(currentPath)) {
+    return false;
+  }
+
+  const isDir = (await fs.stat(currentPath)).isDirectory();
+  // Folder integrity check is simple:
+  // Verify the previous amount of files is not greater than the current one.
+  // We don't want to use equality comparison because of an assumption that the OS might
+  // create some unwanted service files/cached inside of that folder or its subfolders.
+  // Ofc, validating the hash sum of each file (or at least of file path) would be much
+  // more precise, but we don't need to be very precise here and also don't want to
+  // overuse RAM and have a performance drop.
+  if (isDir && await calculateFolderIntegrity(currentPath) >= expectedIntegrity?.folder) {
+    return true;
+  }
+  if (!isDir && await calculateFileIntegrity(currentPath) === expectedIntegrity?.file) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * @typedef {Object} ConfigureAppOptions
+ * @property {?Function} onPostProcess Optional function, which should be applied
+ * to the application after it is downloaded/preprocessed. This function may be async
+ * and is expected to accept single object parameter with the following properties:
+ *
+ * - cachedAppInfo: The information about the previously cached app instance (if exists):
+ *    - packageHash: SHA1 hash of the package if it is a file and not a folder
+ *    - lastModified: Optional Date instance, the value of file's `Last-Modified` header
+ *    - immutable: Optional boolean value. Contains true if the file has an `immutable` mark
+ *                 in `Cache-control` header
+ *    - maxAge: Optional integer representation of `maxAge` parameter in `Cache-control` header
+ *    - timestamp: The timestamp this item has been added to the cache (measured in Unix epoch
+ *                 milliseconds)
+ *    - integrity: An object containing either `file` property with SHA1 hash of the file
+ *                 or `folder` property with total amount of cached files and subfolders
+ *    - fullPath: the full path to the cached app
+ * - isUrl: Whether the app has been downloaded from a remote URL
+ * - headers: Optional headers object. Only present if `isUrl` is true and if the server
+ * responds to HEAD requests.
+ * - appPath: A string containing full path to the preprocessed application package (either
+ * downloaded or a local one)
+ *
+ * The function is expected to either return a falsy value, which means the app must not be
+ * cached and a fresh copy of it is downloaded each time. If this function returns an object
+ * containing `appPath` property then the integrity of it will be verified and stored into
+ * the cache.
+ * @property {!string[]} supportedExtensions List of supported application extensions (
```

Comment:
I'm unsure what the leading `!` is.  does TS understand this?

Reply:
Not sure about that. I was following JSDoc there

Reply:
OK; what it means is `non-null`.  This may make sense in jsdoc, but in typescript, this doesn't make sense (since "non-null" is essentially the default behavior).  imo we should remove `!` and all occurrences thereof

Reply:
I just wanted to emphasise this property is mandatory if an object is passed as an argument. I'm going to mention this in the docstring though